src/Symfony/Component/HttpKernel/ControllerMetadata/ArgumentMetadata.php

```diff
@@ -23,21 +23,28 @@ class ArgumentMetadata
     private $isVariadic;
     private $hasDefaultValue;
     private $defaultValue;
+    private $isNullable;
 
     /**
      * @param string $name
      * @param string $type
      * @param bool   $isVariadic
      * @param bool   $hasDefaultValue
      * @param mixed  $defaultValue
+     * @param bool   $isNullable
      */
-    public function __construct($name, $type, $isVariadic, $hasDefaultValue, $defaultValue)
+    public function __construct($name, $type, $isVariadic, $hasDefaultValue, $defaultValue, $isNullable = null)
     {
         $this->name = $name;
         $this->type = $type;
         $this->isVariadic = $isVariadic;
         $this->hasDefaultValue = $hasDefaultValue;
         $this->defaultValue = $defaultValue;
+        $this->isNullable = (bool) $isNullable;
+
+        if (null === $isNullable) {
+            @trigger_error(sprintf('The $isNullable argument has been added to %s with a default value. As of 4.0 the argument will become mandatory.', __CLASS__), E_USER_DEPRECATED);
```

Comment:
Did a quick test:
https://3v4l.org/cuWM8
https://3v4l.org/QoPoD

It seems like `foo(? string $foo = null)` would work in your assumption but `foo(? string $foo = 'foo')` seems to break. I'm thinking this should be added as test case actually.